File: src/C/linked-list-cycle.c
// https://leetcode.com/problems/linked-list-cycle/

#include "lists.h"


int has_cycle(struct ListNode *head) {
    if (!head) return 0;

    struct ListNode *slow = head;
    struct ListNode *fast = head;
    
    while (fast && fast->next) {
        fast = fast->next->next;
        slow = slow->next;

        if (slow == fast) return 1;
    }

    return 0;
}

File: src/Python/linked-list-cycle.py
#!/usr/bin/env python3
# coding: utf-8
# https://leetcode.com/problems/linked-list-cycle/
class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        while head and not hasattr(head, 'visited'):
            head.visited = True
            head = head.next
        return head != None

